### PR TITLE
Fix annotated_ptr test failures.

### PR DESCRIPTION
**Pull request by wmaxey**
## Description

Fixes build failures encountered by GHA runner bringup.

Annotated pointer tests are currently failing likely due to a pessimism by the compiler. A function that returns shared or global memory based on a runtime value currently breaks `__isshared` assertions.

Though unrelated to CI failures it also fixes some stride loop math.
Closes: #2605

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: libcudacxx/include/cuda/annotated_ptr [hunk 1]
```diff
@@ -268,15 +268,14 @@ apply_access_property(const volatile void* __ptr, const _Shape __shape, access_p
     NV_PROVIDES_SM_80,
     (if (!__isGlobal((void*) __ptr)) return;
 
-     char* __p                               = reinterpret_cast<char*>(const_cast<void*>(__ptr));
-     static constexpr std::size_t _LINE_SIZE = 128;
-     std::size_t __nbytes                    = static_cast<std::size_t>(__shape);
-     std::size_t __end = ((std::uintptr_t)(__p + __nbytes) % _LINE_SIZE) ? __nbytes + _LINE_SIZE : __nbytes;
-     __end /= _LINE_SIZE;
+     uint8_t* __p                   = reinterpret_cast<uint8_t*>(const_cast<void*>(__ptr));
+     constexpr std::size_t __linesz = 128;
```
Comment (miscco) [inline libcudacxx/include/cuda/annotated_ptr:272]: lets just use `__line_size`

File: libcudacxx/test/libcudacxx/cuda/annotated_ptr/utils.h [hunk 2]
```diff
@@ -64,6 +64,18 @@ __device__ __host__ __noinline__ T* alloc(bool shared = false)
...
+
+  for (int i = 0; i < N; ++i)
+  {
+    data[i] = i;
+  }
+  return data;
+}
+
 template <typename T>
 __device__ __host__ __noinline__ void dealloc(T* arr, bool shared)
```
Comment (miscco) [inline libcudacxx/test/libcudacxx/cuda/annotated_ptr/utils.h:80]: Please also drop the `shared` code path from the other allocator

Comment (wmaxey) [reply libcudacxx/test/libcudacxx/cuda/annotated_ptr/utils.h:80]: Could be used elsewhere?

Comment (miscco) [reply libcudacxx/test/libcudacxx/cuda/annotated_ptr/utils.h:80]: I mean it is giving the wrong result so it should not be used elsewhere

Comment (wmaxey) [reply libcudacxx/test/libcudacxx/cuda/annotated_ptr/utils.h]: I made them distinct now.

File: libcudacxx/include/cuda/annotated_ptr [hunk 3]
```diff
@@ -268,15 +268,14 @@ apply_access_property(const volatile void* __ptr, const _Shape __shape, access_p
...
 
-     char* __p                               = reinterpret_cast<char*>(const_cast<void*>(__ptr));
-     static constexpr std::size_t _LINE_SIZE = 128;
-     std::size_t __nbytes                    = static_cast<std::size_t>(__shape);
-     std::size_t __end = ((std::uintptr_t)(__p + __nbytes) % _LINE_SIZE) ? __nbytes + _LINE_SIZE : __nbytes;
-     __end /= _LINE_SIZE;
+     uint8_t* __p                      = reinterpret_cast<uint8_t*>(const_cast<void*>(__ptr));
+     constexpr std::size_t __line_size = 128;
+     const std::size_t __nbytes        = static_cast<std::size_t>(__shape);
+     std::size_t __end                 = 1 + (__nbytes / __line_size);
```
Comment (miscco) [inline libcudacxx/include/cuda/annotated_ptr:274]: should this be `cuda::ceil_div(__nbytes, __line_size)`?

That Happens here if `__nbytes `is divisible by `__line_size`

**Review by miscco — APPROVED** [review]

**Pull request merged**